xlive/Blam/Engine/rasterizer/dx9/rasterizer_dx9_targets.cpp

```diff
@@ -1,22 +1,89 @@
 #include "stdafx.h"
 #include "rasterizer_dx9_targets.h"
-#include "rasterizer/dx9/rasterizer_dx9_main.h"
+
+#include "rasterizer_dx9_main.h"
+
 #include "camera/camera.h"
+#include "main/main_screenshot.h"
+#include "rasterizer/rasterizer_globals.h"
+
+/* constants */
+
+// This is no longer used
+// #define USE_CUBEMAP_TARGETS
+
+/* globals */
+
+real32 g_sun_size = 0.2f;
+
+IDirect3DTexture9* g_backbuffer_texture = NULL;
+// IDirect3DSurface9* g_backbuffer_surface = NULL;
+
+// IDirect3DSurface9* global_d3d_surface_render_primary = NULL;
+// IDirect3DSurface9* global_d3d_surface_render_primary_z = NULL;
+
+// IDirect3DSurface9* global_d3d_surface_render_z_as_target_z = NULL;
+
+IDirect3DTexture9* global_d3d_texture_render_resolved = NULL;
+// IDirect3DSurface9* global_d3d_surface_render_resolved = NULL;
+
+IDirect3DTexture9* global_d3d_texture_sun_glow_primary = NULL;
+// IDirect3DSurface9* global_d3d_surface_sun_glow_primary = NULL;
+IDirect3DTexture9* global_d3d_texture_sun_glow_secondary = NULL;
+IDirect3DSurface9* global_d3d_surface_sun_glow_secondary = NULL;
+
+IDirect3DTexture9* global_d3d_texture_motion_sensor = NULL;
+IDirect3DSurface9* global_d3d_surface_motion_sensor = NULL;
+
+// IDirect3DPixelShader9* global_d3d_shader_white_source = NULL;
+
+bool global_d3d_target_use_depth = false;
 
 /* prototypes */
 
+D3DMULTISAMPLE_TYPE* d3d_multisample_type_get(void);
+uint32* d3d_multisample_quality_get(void);
+
+IDirect3DSurface9** global_d3d_surface_render_z_as_target_z_get(void);
+IDirect3DSurface9** g_backbuffer_surface_get(void);
+IDirect3DPixelShader9** global_d3d_shader_white_source_get(void);
+
+uint32* cubemap_target_size_get(void);
+uint32* motion_sensor_texture_size_get(void);
+IDirect3DTexture9* rasterizer_dx9_get_target_texture(e_rasterizer_target rasterizer_target);
+IDirect3DSurface9* __cdecl rasterizer_dx9_get_target_mip_surface(e_rasterizer_target target, int16 mip);
+void __cdecl rasterizer_set_render_target_internal_hook_set_main_render_surface(IDirect3DSurface9* target, IDirect3DSurface9* z_stencil, bool a3);
+
+bool __cdecl rasterizer_dx9_primary_targets_initialize(void);
+void __cdecl rasterizer_dx9_primary_targets_dispose(void);
+bool __cdecl rasterizer_dx9_secondary_targets_initialize(void);
+void __cdecl rasterizer_dx9_secondary_targets_dispose(void);
 
 
 /* public code */
 
-e_rasterizer_target* rasterizer_dx9_main_render_target_get(void)
+void rasterizer_dx9_targets_apply_patches(void)
 {
-	return Memory::GetAddress<e_rasterizer_target*>(0xA3E228);
+    WriteJmpTo(Memory::GetAddress(0x25FDC0), rasterizer_dx9_set_target);
```

Comment:
`DETOUR_ATTACH` is preferred over `WriteJmpTo`

Reply:
Changed